### PR TITLE
Avoid object creation during each frame

### PR DESCRIPTION
**Pull request by simonla**
`Choreographer.FrameCallback` will  be called each frame

We should avoid create new object here

There is no need to consider thread safety, it is guaranteed  [here](https://github.com/flutter/engine/blob/492cf91f28541be014be308c1e385d5da3cb9f30/shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java#L152) and [here](https://github.com/flutter/engine/blob/266d3360a7babfb5f20d5e9f8ea84772b2a247dc/shell/platform/android/vsync_waiter_android.cc#L44)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by simonla** [comment]
It looks like we will only have one request at same time

So we can only  hold one `FrameCallback` instance

[animator](https://github.com/flutter/engine/blob/main/shell/common/animator.cc#L214)
```cpp
  if (!pending_frame_semaphore_.TryWait()) {
    // Multiple calls to Animator::RequestFrame will still result in a
    // single request to the VsyncWaiter.
    return;
  }

```

File: shell/platform/android/io/flutter/view/VsyncWaiter.java
Comment (jason-simmons) [inline shell/platform/android/io/flutter/view/VsyncWaiter.java:106]: The Java `VsyncWaiter` is a singleton.  But there can be multiple instances of the Flutter engine, each of which has its own `VsyncWaiterAndroid` on the native side.

Each of those `VsyncWaiterAndroid` instances will make JNI calls to `VsyncWaiter.asyncWaitForVsync` and will pass a cookie representing the pointer to the originating `VsyncWaiterAndroid`.

So it is not safe to assume that only one cookie has a pending frame callback.

**Review by jason-simmons — CHANGES_REQUESTED** [review]
Is this causing a significant performance issue?

If the cost of these allocations is noticeable, then one option would be using a collection such as Android's `SparseLongArray` to hold the pending cookies.

File: shell/platform/android/io/flutter/view/VsyncWaiter.java
Comment (simonla) [reply shell/platform/android/io/flutter/view/VsyncWaiter.java:106]: > Is this causing a significant performance issue?
> 
> If the cost of these allocations is noticeable, then one option would be using a collection such as Android's `SparseLongArray` to hold the pending cookies.

You are right. I will collect and dispatch ever single cookie.

Comment (simonla) [reply shell/platform/android/io/flutter/view/VsyncWaiter.java]: I have updated commit, please check

File: shell/platform/android/io/flutter/view/VsyncWaiter.java
Comment (jason-simmons) [inline shell/platform/android/io/flutter/view/VsyncWaiter.java:93]: This constructor can be removed

**Review by jason-simmons — APPROVED** [review]

File: shell/platform/android/io/flutter/view/VsyncWaiter.java
Comment (simonla) [reply shell/platform/android/io/flutter/view/VsyncWaiter.java]: done

**Comment by simonla** [comment]
> Validations Fail.

@jason-simmons help~ 👀

**Review by dnfield — APPROVED** [review]

**Comment by dnfield** [comment]
@simonla this needed a second reviewer, I've approved. Asking for a NTE from @Hixie

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Comment by simonla** [comment]
@jason-simmons Got test-exempt and two approver. Would you mind helping me add the autosubmit tag again？

**Pull request merged**
